File: OPass/Tabs/ScheduleTab/ScheduleTableViewCell.m
//
//  ScheduleTableViewCell.m
//  CCIP
//
//  Created by FrankWu on 2017/7/16.
//  Copyright © 2017年 CPRTeam. All rights reserved.
//

#import "ScheduleTableViewCell.h"
#import "AppDelegate.h"

@interface ScheduleTableViewCell()

@property (readwrite, nonatomic) BOOL favorite;
@property (strong, nonatomic) NSDictionary *schedule;
@property (readwrite, nonatomic) BOOL disabled;

@end

@implementation ScheduleTableViewCell

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
    [self.LabelLabel setTextColor:[UIColor colorFromHtmlColor:@"#9B9B9B"]];
    [self.LabelLabel setBackgroundColor:[UIColor colorFromHtmlColor:@"#D8D8D8"]];
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}

- (NSString *)getID {
    if (self.schedule) {
        if (self.delegate) {
            if ([self.delegate respondsToSelector:@selector(getID:)]) {
                return [self.delegate getID:_schedule];
            }
        }
        return @"";
    } else {
        return @"";
    }
}

- (void)setSchedule:(NSDictionary *)schedule {
    _schedule = schedule;
    NSDate *startTime = [Constants DateFromString:[_schedule objectForKey:@"start"]];
    NSDate *endTime = [Constants DateFromString:[_schedule objectForKey:@"end"]];
    long mins = [[NSNumber numberWithDouble:([endTime timeIntervalSinceDate:startTime] / 60)] longValue];
    NSString *type = [Constants GetScheduleTypeName:[_schedule objectForKey:@"type"]];
    NSDictionary *currentLangObject = [_schedule objectForKey:[AppDelegate shortLangUI]];
    [self.ScheduleTitleLabel setText:[currentLangObject objectForKey:@"title"]];
    [self.RoomLocationLabel setText:[NSString stringWithFormat:@"Room %@ - %ld mins", [_schedule objectForKey:@"room"], mins]];
    [self.LabelLabel setText:[NSString stringWithFormat:@"   %@   ", type]];
    [self.LabelLabel.layer setCornerRadius:self.LabelLabel.frame.size.height / 2];
    [self.LabelLabel sizeToFit];
    [self.LabelLabel setHidden:[type length] == 0];
    [self setFavorite:NO];
    if (self.delegate) {
        if ([self.delegate respondsToSelector:@selector(hasFavorite:)]) {
            [self setFavorite:[self.delegate hasFavorite:[self getID]]];
        }
    }
}

- (NSDictionary *)getSchedule {
    return _schedule;
}

- (IBAction)favoriteTouchDownAction:(id)sender {
    [UIImpactFeedback triggerFeedback:UIImpactFeedbackTypeImpactFeedbackMedium];
}

- (IBAction)favoriteTouchUpInsideAction:(id)sender {
    self.favorite = !self.favorite;
    if (self.delegate) {
        if ([self.delegate respondsToSelector:@selector(actionFavorite:)]) {
            [self.delegate actionFavorite:[self getID]];
        }
    }
    [UIImpactFeedback triggerFeedback:UIImpactFeedbackTypeImpactFeedbackLight];
}

- (IBAction)favoriteTouchUpOutsideAction:(id)sender {
    [UIImpactFeedback triggerFeedback:UIImpactFeedbackTypeImpactFeedbackLight];
}

- (void)setFavorite:(BOOL)favorite {
    _favorite = favorite;
    NSDictionary *titleAttribute = @{
                                     NSFontAttributeName: [Constants fontOfAwesomeWithSize:20 inStyle:(_favorite ? fontAwesomeStyleSolid : fontAwesomeStyleRegular)],
                                     NSForegroundColorAttributeName: [AppDelegate AppConfigColor:@"FavoriteButtonColor"],
                                     };
    NSAttributedString *title = [[NSAttributedString alloc] initWithString:[Constants fontAwesomeWithCode:@"fa-heart"] attributes:titleAttribute];
    [self.FavoriteButton setAttributedTitle:title
                                   forState:UIControlStateNormal];
}

- (BOOL)getFavorite {
    return _favorite;
}

- (void)setDisabled:(BOOL)disabled {
    _disabled = disabled;
    [self.ScheduleTitleLabel setAlpha:_disabled ? .2f : 1];
}

- (BOOL)getDisabled {
    return _disabled;
}

@end

File: OPass/Tabs/ScheduleTab/ScheduleTableViewCell.swift
//
//  ScheduleTableViewCell.swift
//  OPass
//
//  Created by 腹黒い茶 on 2019/6/9.
//  Copyright © 2019 OPass. All rights reserved.
//

import Foundation
import UIKit

class ScheduleTableViewCell: UITableViewCell {
    public var delegate: ScheduleFavoriteDelegate?
    @IBOutlet public var ScheduleTitleLabel: UILabel?
    @IBOutlet public var RoomLocationLabel: UILabel?
    @IBOutlet public var LabelLabel: UILabel?
    @IBOutlet public var FavoriteButton: UIButton?

    private var favorite: Bool = false
    private var disabled: Bool = false
    private var session: SessionInfo?

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
        self.LabelLabel?.textColor = UIColor.colorFromHtmlColor("#9b9b9b")
        self.LabelLabel?.backgroundColor = UIColor.colorFromHtmlColor("#d8d8d8")
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)
        // Configure the view for the selected state
    }

    @IBAction func favoriteTouchDownAction(_ sender: NSObject) {
        UIImpactFeedback.triggerFeedback(.impactFeedbackMedium)
    }

    @IBAction func favoriteTouchUpInsideAction(_ sender: NSObject) {
        self.favorite = !self.favorite
        if (self.delegate != nil) {
            if ((self.delegate?.responds(to: #selector(ScheduleFavoriteDelegate.actionFavorite(_:))))!) {
                self.delegate!.actionFavorite(self.getID())
            }
        }
        UIImpactFeedback.triggerFeedback(.impactFeedbackLight)
    }

    @IBAction func favoriteTouchUpOutsideAction(_ sender: NSObject) {
        UIImpactFeedback.triggerFeedback(.impactFeedbackLight)
    }

    func getID() -> String {
        return self.session?.Id ?? ""
    }

    func setSession(_ session: SessionInfo) {
        self.session = session

        let startTime = Constants.DateFromString(self.session!.Start)
        let endTime = Constants.DateFromString(self.session!.End)
        let mins = Int(endTime.timeIntervalSince(startTime) / 60)
        self.RoomLocationLabel?.text = "Room \(self.session!.Room!) - \(mins) mins"

        self.ScheduleTitleLabel?.text = self.session!["title"]

        let type = Constants.GetScheduleTypeName(self.session!.Type!)
        self.LabelLabel?.text = "   \(type)   "
        self.LabelLabel?.layer.cornerRadius = (self.LabelLabel?.frame.size.height)! / 2
        self.LabelLabel?.sizeToFit()
        self.LabelLabel?.isHidden = type.count == 0
        self.setFavorite(false)

        if (self.delegate != nil) {
            if ((self.delegate?.responds(to: #selector(ScheduleFavoriteDelegate.hasFavorite(_:))))!) {
                self.setFavorite(self.delegate!.hasFavorite(self.getID()))
            }
        }
    }

    func getSession() -> SessionInfo? {
        return self.session
    }

    @objc func setFavorite(_ favorite: Bool) {
        self.favorite = favorite
        let titleAttribute = [
            NSAttributedString.Key.font: Constants.fontOfAwesome(withSize: 20, inStyle: self.favorite ? fontAwesomeStyle.solid : fontAwesomeStyle.regular),
            NSAttributedString.Key.foregroundColor: AppDelegate.appConfigColor("FavoriteButtonColor")
        ]
        let title = NSAttributedString.init(string: Constants.fontAwesome(code: "fa-heart")!, attributes: titleAttribute)
        self.FavoriteButton?.setAttributedTitle(title, for: .normal)
    }

    @objc func getFavorite() -> Bool {
        return self.favorite
    }

    @objc func setDisabled(_ disabled: Bool) {
        self.disabled = disabled
        self.ScheduleTitleLabel?.alpha = self.disabled ? 0.2 : 1
    }

    @objc func getDisabled() -> Bool {
        return self.disabled
    }

    @objc func setDelegate(_ delegate: ScheduleFavoriteDelegate) {
        self.delegate = delegate
    }
}
